contracts/secretswap_router/src/contract.rs

```diff
@@ -0,0 +1,378 @@
+use cosmwasm_std::{
+    debug_print, from_binary, to_binary, Api, Binary, Coin, CosmosMsg, Env, Extern, HandleResponse,
+    InitResponse, Querier, StdError, StdResult, Storage, WasmMsg,
+};
+use secret_toolkit::snip20;
+use secretswap::{Asset, AssetInfo};
+use HandleMsg::RecoverFunds;
+
+use crate::{
+    msg::{HandleMsg, Hop, InitMsg, NativeSwap, QueryMsg, Route, Snip20Swap},
+    state::{
+        delete_route_state, read_owner, read_route_state, read_tokens, store_owner,
+        store_route_state, store_tokens, RouteState,
+    },
+};
+
+pub fn init<S: Storage, A: Api, Q: Querier>(
+    deps: &mut Extern<S, A, Q>,
+    env: Env,
+    _msg: InitMsg,
+) -> StdResult<InitResponse> {
+    store_owner(&mut deps.storage, &env.message.sender)?;
+    store_tokens(&mut deps.storage, &vec![])?;
+
+    Ok(InitResponse::default())
+}
+
+pub fn handle<S: Storage, A: Api, Q: Querier>(
+    deps: &mut Extern<S, A, Q>,
+    env: Env,
+    msg: HandleMsg,
+) -> StdResult<HandleResponse> {
+    match msg {
+        HandleMsg::Receive {
```

Comment:
Do we want to create a specific message for receiving & swapping uscrt? The first message could be a call to the Receive function or something

Reply:
I don't think that's necessary. And I think the code will be a bit more branched out on both contract and client side.
Please take a look at the code now and tell me I you think that's necessary.